File: mac_app/Services/Translation/ProxyTranslationProvider.swift
import Foundation

/// Routes translation requests through the QueenMama backend (`/api/proxy/translate`).
/// Mirrors the philosophy of `ProxyAIProvider`: no user-managed API key, billing centralized,
/// auth via existing bearer token.
final class ProxyTranslationProvider: TranslationProvider {
    let providerName = "DeepL (Proxy)"

    init() {}

    var isConfigured: Bool {
        MainActor.assumeIsolated {
            ProxyConfigManager.shared.isTranslationEnabled
        }
    }

    func translate(text: String, sourceLang: String?, targetLang: String) async throws -> TranslationResult {
        let start = Date()
        do {
            let response = try await ProxyAPIClient.shared.translate(
                text: text,
                sourceLang: sourceLang,
                targetLang: targetLang
            )
            let latency = Int(Date().timeIntervalSince(start) * 1000)
            return TranslationResult(
                translatedText: response.translatedText,
                detectedSourceLang: response.detectedSourceLang,
                targetLang: response.targetLang ?? targetLang,
                provider: providerName,
                latencyMs: latency,
                cached: false
            )
        } catch let error as ProxyError {
            switch error {
            case .notAuthenticated:
                throw TranslationError.notConfigured
            case .requestFailed(let code, _) where code == 429:
                throw TranslationError.rateLimited
            case .requestFailed(let code, let msg) where code == 402:
                throw TranslationError.quotaExceeded
            default:
                throw TranslationError.requestFailed(error.localizedDescription)
            }
        } catch {
            throw TranslationError.requestFailed(error.localizedDescription)
        }
    }
}

File: win_app_2/src/services/translation/proxyTranslationProvider.ts
import { translate as proxyTranslate } from '@/services/proxy/proxyApiClient'
import { isTranslationEnabled } from '@/services/proxy/proxyConfigManager'
import {
  TranslationError,
  type TranslationProvider,
  type TranslationResult,
} from './translationProvider'

export class ProxyTranslationProvider implements TranslationProvider {
  readonly providerName = 'DeepL (Proxy)'

  get isConfigured(): boolean {
    return isTranslationEnabled()
  }

  async translate(input: {
    text: string
    sourceLang?: string | null
    targetLang: string
    context?: string | null
  }): Promise<TranslationResult> {
    const start = Date.now()
    try {
      const response = await proxyTranslate({
        text: input.text,
        sourceLang: input.sourceLang ?? undefined,
        targetLang: input.targetLang,
        context: input.context ?? undefined,
      })
      return {
        translatedText: response.translated_text,
        detectedSourceLang: response.detected_source_lang ?? null,
        targetLang: response.target_lang ?? input.targetLang,
        provider: this.providerName,
        latencyMs: Date.now() - start,
        cached: false,
      }
    } catch (err) {
      if (err instanceof TranslationError) throw err
      const message = err instanceof Error ? err.message : String(err)
      throw new TranslationError('requestFailed', message)
    }
  }
}
